xtevent/_eventgenvars.ado

```diff
@@ -1,306 +1,499 @@
+		
+		cap by `panelvar' (`timevar'): gen `l1'= (F1.`zr'>=`zr') if !missing(`zr') & !missing(F1.`zr') & `touse'
+		cap assert `l1'==1 if !missing(`l1')
+		if ! _rc{
+			loc norever 1
+		}
+	}
+	
+	if `norever'==0 & ("`impute'"=="stag" | "`impute'"=="instag") {
+		di "Policyvar changes more than once for some units. Assuming non-staggered adoption (no imputation)."
+		loc impute=""
+	}
+
+	
+	****** if no-reversion holds, verify "bounds" condition: e.g. if binary 0 and 1, verify 0 as the first observed value and 1 as the last observed value
+	
+	loc bounds 0
+	if "`impute'"=="stag" | "`impute'"=="instag" {
+		tempvar notmiss zt minzt maxzt
+		cap gen `notmiss'=!missing(`z')
+		
+		cap by `panelvar' (`timevar'): gen `zt'=`timevar' if `notmiss'==1 
+		cap by `panelvar' (`timevar'): egen `maxzt'=max(`zt') 
+		cap by `panelvar' (`timevar'): egen `minzt'=min(`zt')
+
+		*first filter: all units satisfy the bounds condition? 
+		if `bin'==1 & `norever'==1 {
+			*verify the lower-bound value
+			cap assert `z'==`rminz' if `minzt'==`timevar' & `touse'
+			* if the lower-bound value is zero, then test the upper-bound
+			if !_rc{
+				cap assert `z'==`rmaxz' if `maxzt'==`timevar' & `touse'
+				if !_rc loc bounds 1
+			} 
+		}
+		
+		*second filter: if some units didn't satisfy the first filter, then allow those units to have only adopted policy or only unadopted policy (and no missing values inside the observed range)
+		if `bin'==1 & `norever'==1 & `bounds'==0 {
+		
+		tempvar ilb iub sb sbmin
+		cap by `panelvar' (`timevar'): gen `ilb'=(`z'==`rminz') if `minzt'==`timevar' & `touse'
+		cap by `panelvar' (`timevar'): gen `iub'=(`z'==`rmaxz') if `maxzt'==`timevar' & `touse'
+		cap egen `sb'=rowtotal(`ilb' `iub') if (`minzt'==`timevar' | `maxzt'==`timevar')
+		*sbmin is an indicator of the units that satisfied the first filter 
+		cap by `panelvar' (`timevar'): egen `sbmin'=min(`sb') if `timevar'>=`zmint2' & `timevar'<=`zmaxt2' & `touse'
+		
+		cap assert inlist(`z',`rminz',`rmaxz') if `sbmin'==0
+		if !_rc loc bounds 1
+		}
+	}
+	
+	if `bounds'==0 & ("`impute'"=="stag" | "`impute'"=="instag") {
+			di "For some units, the changes in policyvar are not consistent with no-unobserved-change. Reverting to default (no imputation)."
+			loc impute =""	
+	}
+	***************** no unobserved change ***************************
+
+	if "`impute'"!="" {
+		cap replace zn2=`zminv2' if `timevar'<`zmint2'
+		cap replace zn2=`zmaxv2' if `timevar'>`zmaxt2'
+	}
+
+	************** impute inner missing values ***********************
+	if "`impute'"=="instag" {
+		
+		tempvar zdown zup 
+		cap gen `zdown'=`z'
+		gsort `panelvar' `timevar'
+		cap replace `zdown'=`zdown'[_n-1] if missing(`zdown')
+		cap gen `zup'=`z'
+		gsort + `panelvar' - `timevar'
+		cap replace `zup'=`zup'[_n-1] if missing(`zup')
+		sort `panelvar' `timevar'
+		
+		cap replace zn2=`zdown' if `timevar'>=`zmint2' & `timevar'<=`zmaxt2' & missing(`z') & `zdown'==`zup'

```

Comment:
By using this [**filter**](https://github.com/JMSLab/xtevent/blob/0d383ff5b427eb4a0b79061914e7f764ac0fc53d/xtevent/_eventgenvars.ado#L203), `impute(stag)` and `impute(instag)` will also allow cases where the policy variable has only adopted policy state values or only unadopted policy state values, but there must not be missing values inside the observed data range. Thus, it will not performe imputation in cases as the one pointed out.

Reply:
@Constantino-Carreto-Romero I am not following here. Should `impute(instag)` impute zeros in the middle in this scenario?

Reply:
If almost all units satisfy the staggered conditions (binary and no-reversion) and the extra condition ("bounds" condition) from issue #15, but there is at least one unit that satisfied the staggered conditions but not the bounds condition, then `impute(stag)` and `impute(instag)` would revert to no imputation. Nonetheless, if these units that didn't satisfy the bounds condition (for which imputation would fail) were only zeros or only ones (if binary of type 0 and 1, for instance), with no missing values inside the observed data range, like this one: 
|time|z|
| --- | --- |
|1|.|
|2|0|
|3|0|
|4|0|
|5|0|
|6|.|

Then these units, although not satisfying the bounds condition as defined, would be satisfying the staggered conditions and we shouldn't let `impute(stag)` and `impute(instag)` fail due to these units. Therefore, we are letting units of this kind satisfy the bounds conditions as well, by using [this block of code](https://github.com/JMSLab/xtevent/blob/0d383ff5b427eb4a0b79061914e7f764ac0fc53d/xtevent/_eventgenvars.ado#L203). On the contrary, units with only zeros or ones, but with missing values inside the observed data range, like this one:
|time|z|
| --- | --- |
|1|.|
|2|0|
|3|.|
|4|.|
|5|0|
|6|.|

will cause `impute(stag)` and `impute(instag)` revert to no imputation.

Reply:
@veli-m-andirin Let me know if this makes sense to resolve the issue

Reply:
Thanks @Constantino-Carreto-Romero @jorpppp, this makes sense as well.